File: ios/Classes/Methods/MessagesMethods.swift
import Flutter
import TwilioConversationsClient

public class MessagesMethods {
    public static func sendMessage(_ call: FlutterMethodCall, _ flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
            let conversationSid = arguments["conversationSid"] as? String,
            let options = arguments["options"] as? [String: Any?] else {
                return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }
        
        let messageOptions = TCHMessageOptions()
        
        if let messageBody = options["body"] as? String {
            messageOptions.withBody(messageBody)
        }
        if let input = options["inputPath"] as? String {
            guard let mimeType = options["mimeType"] as? String else {
                return flutterResult(FlutterError(code: "ERROR", message: "Missing 'mimeType' in MessageOptions", details: nil))
            }
            
            if let inputStream = InputStream(fileAtPath: input) {
                messageOptions.withMediaStream(inputStream, contentType: mimeType, defaultFilename: options["filename"] as? String,
                                               onStarted: {
                    // TODO
                                                
                },
                                               onProgress: { (bytes: UInt) in
                    // TODO
                                                
                },
                                               onCompleted: { (mediaSid: String) in
                    // TODO
                                                
                })
            } else {
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving file for upload from `\(input)`", details: nil))
            }
        }
        
        
        SwiftTwilioConversationsPlugin.instance?.client?.conversation(withSidOrUniqueName: conversationSid, completion: { (result: TCHResult, conversation: TCHConversation?) in
            if result.isSuccessful,
               let conversation = conversation {
                conversation.sendMessage(with: messageOptions, completion: { (result: TCHResult, message: TCHMessage?) in
                    if result.isSuccessful,
                       let message = message {
                        SwiftTwilioConversationsPlugin.debug("\(call.method) (Message.sendMessage) => onSuccess")
                        flutterResult(Mapper.encode(Mapper.messageToDict(message, conversationSid: conversationSid)))
                    } else {
                        SwiftTwilioConversationsPlugin.debug("\(call.method) (Message.sendMessage) => onError: \(result.error.debugDescription)")
                        flutterResult(FlutterError(code: "ERROR", message: "Error sending message with options `\(String(describing: messageOptions))`", details: nil))
                    }
                })
            }
        })
    }

    public static func getMessagesBefore(_ call: FlutterMethodCall, _ flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
              let conversationSid = arguments["conversationSid"] as? String,
              let index = arguments["index"] as? UInt,
              let count = arguments["count"] as? UInt else {
            return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }
        
        SwiftTwilioConversationsPlugin.instance?.client?.conversation(withSidOrUniqueName: conversationSid, completion: { (result: TCHResult, conversation: TCHConversation?) in
            if result.isSuccessful, let conversation = conversation {
                conversation.getMessagesBefore(index, withCount: count, completion: { (result: TCHResult, messages: [TCHMessage]?) in
                    if result.isSuccessful, let messages = messages {
                        SwiftTwilioConversationsPlugin.debug("\(call.method) (Messages.getBefore) => onSuccess")
                        let messagesMap = messages.map { message in
                            Mapper.messageToDict(message, conversationSid: conversationSid)
                        }
                        flutterResult(Mapper.encode(messagesMap))
                    } else {
                        SwiftTwilioConversationsPlugin.debug("\(call.method) (Messages.getBefore) => onError: \(String(describing: result.error))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error retrieving \(count) messages before message (index: \(index)) from conversation (sid: \(conversationSid))", details: nil))
                    }
                })
            }
        })
    }
    
    public static func getLastMessages(_ call: FlutterMethodCall, _ flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
              let count = arguments["count"] as? UInt,
              let conversationSid = arguments["conversationSid"] as? String else {
            return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }
        
        SwiftTwilioConversationsPlugin.instance?.client?.conversation(withSidOrUniqueName: conversationSid, completion: { (result: TCHResult, conversation: TCHConversation?) in
            if result.isSuccessful, let conversation = conversation {
                conversation.getLastMessages(withCount: count, completion: { (result: TCHResult, messages: [TCHMessage]?) in
                    if result.isSuccessful, let messages = messages {
                        SwiftTwilioConversationsPlugin.debug("\(call.method) => onSuccess")
                        let messagesMap = messages.map { message in
                            Mapper.messageToDict(message, conversationSid: conversationSid)
                        }
                        flutterResult(Mapper.encode(messagesMap))
                    } else {
                        flutterResult(FlutterError(code: "ERROR", message: "Error retrieving last \(count) messages for conversation with sid '\(conversationSid)'", details: nil))
                    }
                })
            } else {
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving conversation with sid '\(conversationSid)'", details: nil))
            }
        })
    }
    
    public static func setAllMessagesRead(_ call: FlutterMethodCall, _ flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
              let conversationSid = arguments["conversationSid"] as? String else {
            return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }
        
        SwiftTwilioConversationsPlugin.instance?.client?.conversation(withSidOrUniqueName: conversationSid, completion: { (result: TCHResult, conversation: TCHConversation?) in
            if result.isSuccessful,
               let conversation = conversation {
                conversation.setAllMessagesReadWithCompletion({ (result: TCHResult, count: UInt) in
                    if result.isSuccessful {
                        SwiftTwilioConversationsPlugin.debug("\(call.method) => onSuccess")
                        flutterResult(count)
                    } else {
                        SwiftTwilioConversationsPlugin.debug("\(call.method) => onError: \(String(describing: result.error))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error setting all messages read for conversation (sid: \(conversationSid))", details: nil))
                    }
                })
            } else {
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving conversation with sid '\(conversationSid)'", details: nil))
            }
        })
    }
    
    public static func setLastReadMessageIndex(_ call: FlutterMethodCall, _ flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
              let conversationSid = arguments["conversationSid"] as? String,
              let lastReadMessageIndex = arguments["lastReadMessageIndex"] as? NSNumber else {
            return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }
        
        SwiftTwilioConversationsPlugin.instance?.client?.conversation(withSidOrUniqueName: conversationSid, completion: { (result: TCHResult, conversation: TCHConversation?) in
            if result.isSuccessful,
               let conversation = conversation {
                conversation.setLastReadMessageIndex(lastReadMessageIndex, completion: { (result: TCHResult, count: UInt) in
                    if result.isSuccessful {
                        SwiftTwilioConversationsPlugin.debug("\(call.method) => onSuccess")
                        flutterResult(count)
                    } else {
                        SwiftTwilioConversationsPlugin.debug("\(call.method) => onError: \(result.error.debugDescription))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error setting last consumed message index (index: \(lastReadMessageIndex)) for conversation (sid: \(conversationSid))", details: nil))
                    }
                })
            } else {
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving conversation with sid '\(conversationSid)'", details: nil))
            }
        })
    }
}

File: android/src/main/kotlin/twilio/flutter/twilio_conversations/methods/MessagesMethods.kt
package twilio.flutter.twilio_conversations.methods

import com.google.gson.Gson
import com.twilio.conversations.*
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import twilio.flutter.twilio_conversations.Mapper
import twilio.flutter.twilio_conversations.TwilioConversationsPlugin
import java.io.FileInputStream


object MessagesMethods {
    fun getLastMessages(call: MethodCall, result: MethodChannel.Result) {
        val count = call.argument<Int>("count")
                ?: return result.error("ERROR", "Missing 'count'", null)

        val conversationSid = call.argument<String>("conversationSid")
                ?: return result.error("ERROR", "Missing 'conversationSid'", null)

        try {
            TwilioConversationsPlugin.client?.getConversation(conversationSid, object : CallbackListener<Conversation> {
                override fun onSuccess(conversation: Conversation?) {
                    conversation?.getLastMessages(count, object : CallbackListener<List<Message>> {
                        override fun 
                                onSuccess(messages: List<Message>) {
                            TwilioConversationsPlugin.debug("${call.method} => onSuccess")

                            val jsonMap = Gson().toJson(messages.map { Mapper.messageToMap(it) })
                            result.success(jsonMap)
                        }

                        override fun onError(errorInfo: ErrorInfo) {
                            TwilioConversationsPlugin.debug("${call.method} (conversation.getLastMessages) => onError: $errorInfo")
                            result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                        }
                    })
                }

                override fun onError(errorInfo: ErrorInfo) {
                    TwilioConversationsPlugin.debug("${call.method} => onError: $errorInfo")
                    result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                }
            })
        } catch (err: IllegalArgumentException) {
            return result.error("IllegalArgumentException", err.message, null)
        } catch (err: IllegalStateException) {
            return result.error("IllegalArgumentException", err.message, null)
        }
    }
    fun sendMessage(call: MethodCall, result: MethodChannel.Result) {
        val options = call.argument<Map<String, Any>>("options")
                ?: return result.error("ERROR", "Missing 'options'", null)

        val conversationSid = call.argument<String>("conversationSid")
                ?: return result.error("ERROR", "Missing 'conversationSid'", null)

        val messageOptions = Message.options()
        if (options["body"] != null) {
            messageOptions.withBody(options["body"] as String)
        }

        if (options["inputPath"] != null) {
            val input = options["inputPath"] as String
            val mimeType = options["mimeType"] as String?
                    ?: return result.error("ERROR", "Missing 'mimeType' in MessageOptions", null)

            messageOptions.withMedia(FileInputStream(input), mimeType)
            if (options["filename"] != null) {
                messageOptions.withMediaFileName(options["filename"] as String)
            }
        }

        try {
            TwilioConversationsPlugin.client?.getConversation(conversationSid, object : CallbackListener<Conversation> {
                override fun onSuccess(conversation: Conversation) {
                    conversation.sendMessage(messageOptions) { message ->
                        TwilioConversationsPlugin.debug("${call.method} => onSuccess")

                        val jsonMap = Gson().toJson(Mapper.messageToMap(message))
                        result.success(jsonMap)
                    }
                }

                override fun onError(errorInfo: ErrorInfo) {
                    TwilioConversationsPlugin.debug("${call.method} => onError: $errorInfo")
                    result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                }
            })
        } catch (err: IllegalArgumentException) {
            return result.error("IllegalArgumentException", err.message, null)
        }
    }

    fun getMessagesBefore(call: MethodCall, result: MethodChannel.Result) {
        val index = call.argument<Int>("index")?.toLong()
                ?: return result.error("ERROR", "Missing 'index'", null)
        val count = call.argument<Int>("count")
                ?: return result.error("ERROR", "Missing 'count'", null)
        val conversationSid = call.argument<String>("conversationSid")
                ?: return result.error("ERROR", "Missing 'conversationSid'", null)

        TwilioConversationsPlugin.client?.getConversation(conversationSid, object : CallbackListener<Conversation> {
            override fun onSuccess(conversation: Conversation) {
                conversation.getMessagesBefore(index, count, object : CallbackListener<List<Message>> {
                    override fun onSuccess(messages: List<Message>) {
                        val jsonMap = Gson().toJson(messages.map { Mapper.messageToMap(it) })
                        result.success(jsonMap)
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        TwilioConversationsPlugin.debug("${call.method} (conversation.getMessagesBefore) => onError: $errorInfo")
                        result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioConversationsPlugin.debug("${call.method} => onError: $errorInfo")

                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun setAllMessagesRead(call: MethodCall, result: MethodChannel.Result) {
        val conversationSid = call.argument<String>("conversationSid")
                ?: return result.error("ERROR", "Missing 'conversationSid'", null)

        TwilioConversationsPlugin.client?.getConversation(conversationSid, object : CallbackListener<Conversation> {
            override fun onSuccess(conversation: Conversation) {
                conversation.setAllMessagesRead(object : CallbackListener<Long> {
                    override fun onSuccess(index: Long) {
                        result.success(index)
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        TwilioConversationsPlugin.debug("${call.method} (conversation.setAllMessagesRead) => onError: $errorInfo")
                        result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioConversationsPlugin.debug("${call.method} => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun setLastReadMessageIndex(call: MethodCall, result: MethodChannel.Result) {
        val conversationSid = call.argument<String>("conversationSid")
                ?: return result.error("ERROR", "Missing 'conversationSid'", null)
        val lastReadMessageIndex = call.argument<Int>("lastReadMessageIndex")?.toLong()
                ?: return result.error("ERROR", "Missing 'lastReadMessageIndex'", null)

        TwilioConversationsPlugin.client?.getConversation(conversationSid, object : CallbackListener<Conversation> {
            override fun onSuccess(conversation: Conversation) {
                conversation.setLastReadMessageIndex(lastReadMessageIndex, object : CallbackListener<Long> {
                    override fun onSuccess(newIndex: Long) {
                        result.success(newIndex)
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        TwilioConversationsPlugin.debug("${call.method} (conversation.setLastReadMessageIndex) => onError: $errorInfo")
                        result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioConversationsPlugin.debug("${call.method} => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }
}
